pylearn2/linear/cudnn2d.py

```diff
@@ -0,0 +1,162 @@
+import functools
+import numpy as np
+
+from theano.sandbox.cuda.dnn import GpuDnnConv, GpuDnnConvDesc
+from theano.sandbox.cuda.basic_ops import gpu_contiguous
+
+from pylearn2.packaged_dependencies.theano_linear.conv2d \
+    import Conv2d as OrigConv2D
+
+from pylearn2.linear.linear_transform import LinearTransform as P2LT
+from pylearn2.utils import sharedX
+from pylearn2.utils.rng import make_np_rng
+
+
+default_seed = [2012, 11, 6, 9]
+default_sparse_seed = [2012, 11, 6]
+
+
+class Cudnn2D(OrigConv2D):
+    """
+    Wrapper on the Theano Cudnn op.
+    """
+
+    def __init__(self, filters,
+                 batch_size,
+                 input_space,
+                 output_axes=('b', 'c', 0, 1),
+                 subsample=(1, 1),
+                 pad=(0, 0),
+                 border_mode='valid',
+                 filters_shape=None,
+                 message=''):
+
+        assert batch_size is None or batch_size > 0
+        self.input_space = input_space
+        self.output_axes = output_axes
+        self._pad = pad
+        self.subsample = tuple(subsample)
+
+        super(Cudnn2D, self).__init__(
+            filters=filters,
+            img_shape=(batch_size, input_space.num_channels,
+                       input_space.shape[0], input_space.shape[1]),
+            subsample=self.subsample,
+            border_mode=border_mode,
+            filters_shape=filters.get_value(borrow=True).shape,
+            message=message
+        )
+
+        # conv_op has to be changed
+        self.conv_op = GpuDnnConv()
+        self.desc = GpuDnnConvDesc(border_mode=border_mode,
+                                   subsample=self.subsample,
+                                   conv_mode='conv')
+        # self.conv_op = GpuCorrMM(subsample=self._subsample,
+        #                         border_mode=border_mode,
+        #                         pad=pad)
+
+    @functools.wraps(P2LT.get_params)
+    def get_params(self):
+        """
+        .. todo::
+
+            WRITEME
+        """
+        return [self._filters]
+
+    @functools.wraps(P2LT.get_weights_topo)
+    def get_weights_topo(self, borrow):
+        """
+        .. todo::
+
+            WRITEME
+        """
+        return np.transpose(self._filters.get_value(borrow=borrow),
+                            (0, 2, 3, 1))
+
+    def lmul(self, x):
+        """
+        .. todo::
+
+            WRITEME properly
+
+        dot(x, A)
+
+        This method overrides the original CorrMM2D lmul to make it work
+        with arbitrary axis orders
+        """
+
+        # x must be formatted as batch index, channel, topo dim 0, topo dim 1
+        # for use with conv2d, so check what the current input space format is
+        assert x.ndim == 4
+        axes = self.input_space.axes
+        assert len(axes) == 4
+
+        op_axes = ('b', 'c', 0, 1)
+
+        if tuple(axes) != op_axes:
+            x = x.dimshuffle(
```

Comment:
I think you could write this as
```
x = x.dimshuffle(*[axes.index(ax) for ax in op_axes])
```
and it would be more concise and reuse the ```op_axes``` definition.